In Memory Bridge Response

EasterTheBunny opened the pull request:
This commit adds a cache to the bridges ORM for responses and upserts responses async instead of waiting for a database call to complete. Get requests for responses are lazy in that they first look in the memory cache before querying the database.

samsondav commented on core/bridges/orm.go:
```diff
@@ -179,26 +185,43 @@ func (o *orm) UpdateBridgeType(ctx context.Context, bt *BridgeType, btr *BridgeT
-func (o *orm) GetCachedResponse(ctx context.Context, dotId string, specId int32, maxElapsed time.Duration) (response []byte, err error) {
+func (o *orm) GetCachedResponse(ctx context.Context, dotId string, specId int32, maxElapsed time.Duration) ([]byte, error) {
+	// prefer to get latest value from cache
+	cachedResponse, inCache := o.bridgeLastValueCache.Load(responseKey(dotId, specId))
```
How do we manage the case where the response is expired?

EasterTheBunny replied:
I could store the time coming in and going out and return an error when `maxElapsed` is less than or equal to the stored time value. I started to do that, but left it out to keep the original query in-tact.

samsondav replied:
Why not persist the value as a struct like `{val, time}` then you can check to see if its expired on lookup as well as run a background cleanup routine for pruning stale values.

samsondav commented on core/bridges/orm.go:
```diff
@@ -179,26 +185,43 @@ func (o *orm) UpdateBridgeType(ctx context.Context, bt *BridgeType, btr *BridgeT
-	_, err := o.ds.ExecContext(ctx, sql, dotId, specId, response, time.Now())
-	return pkgerrors.Wrap(err, "failed to upsert bridge response")
+	// async call to the data store with retry
+	go o.upsertResponse(ctx, dotId, specId, response, UpsertRetryDelay)
```
Problem here is that we will try to open an additional DB connection for every stored result, and the number of connections is starting to become an issue. Could we consider instead feeding responses into a queue which is pulled from on a loop that does multiple inserts at a time, so it only consumes one connection?

You could also optimize so that only the latest for every key is persisted.

EasterTheBunny replied:
Ah I see. I was thinking the problem was only latency, but if we're having a connection pool problem too, bulk inserts are the way to go. However, how much of this data is necessary to keep? Doesn't the app only want the latest one anyway? And one part of the problem is pruning unnecessary data from the database, resulting in table locks and latency?

samsondav replied:
Yes we only care about latest, so your loop will take the map of latest values for every key on each iteration and persist all those together. You can run a background pruning loop to dump stale values.

samsondav commented on core/bridges/orm.go:
```diff
@@ -266,3 +289,28 @@ func (o *orm) FindExternalInitiatorByName(ctx context.Context, iname string) (ex
+
+	_, err := o.ds.ExecContext(ctx, sql, dotId, specId, response, time.Now())
+	if err != nil {
+		// TODO: log a warning and retry
+		time.Sleep(nextDelay)
```
This can lead to resource exhaustion as implemented

EasterTheBunny replied:
How quickly are these records being saved? Or better, how many records per 10s should we be writing to the database?

samsondav replied:
Depends on the product, mercury will do hundreds per second or more

samsondav replied:
If you get a persistent error e.g. some constraint is getting violated or DB is unavailable this would spawn unbounded goroutines as written

EasterTheBunny committed:
In Memory Bridge Response

This commit adds a cache to the bridges ORM for responses and upserts responses async instead
of waiting for a database call to complete. Get requests for responses are lazy in that they
first look in the memory cache before querying the database.

EasterTheBunny committed:
initialize internal values

EasterTheBunny committed:
add tests and refactor constructor

EasterTheBunny committed:
fix orm query

jmank88 commented on core/services/pipeline/runner.go:
```diff
@@ -156,7 +181,15 @@ func (r *runner) Name() string {
 
 func (r *runner) HealthReport() map[string]error {
-	return map[string]error{r.Name(): r.Healthy()}
+	cache, isCache := r.btORM.(*bridges.BridgeCache)
```
Could we just check for the HealthReporter interface?

EasterTheBunny replied:
I changed the check, but it still just checks `Ready()`. I would like to run `HealthReport()` but that returns a map and I don't know a clean way to combine maps other than just looping through both.

jmank88 replied:
`services.CopyHealth` is typically used https://github.com/smartcontractkit/chainlink-common/blob/main/pkg%2Fservices%2Fhealth.go#L18

jmank88 commented on core/services/pipeline/task.bridge_test.go:
```diff
@@ -1069,7 +1069,15 @@ func TestBridgeTask_AdapterResponseStatusFailure(t *testing.T) {
+	// orm := bridges.NewORM(db)
+	orm := bridges.NewBridgeCache(bridges.NewORM(db), nil, bridges.DefaultUpsertInterval)
+
+	t.Cleanup(func() {
+		require.NoError(t, orm.Close())
+	})
+
+	require.NoError(t, orm.Start(context.Background()))
```
```suggestion
	servicetest.Run(t, orm)
```

jmank88 commented on core/services/pipeline/runner.go:
```diff
@@ -147,6 +166,12 @@ func (r *runner) Close() error {
 		close(r.chStop)
 		r.wgDone.Wait()
+
+		cache, isCache := r.btORM.(*bridges.BridgeCache)
```
```suggestion
		closer, ok := r.btORM.(io.Closer)
```

jmank88 commented on core/gethwrappers/generated/automation_registry_logic_c_wrapper_2_3/automation_registry_logic_c_wrapper_2_3.go:
What produced this?

EasterTheBunny replied:
I just ran `make generate`.

jmank88 commented on core/bridges/orm.go:
```diff
@@ -198,7 +191,36 @@ func (o *orm) UpsertBridgeResponse(ctx context.Context, dotId string, specId int
+func (o *orm) BulkUpsertBridgeResponse(ctx context.Context, responses []BridgeResponse) error {
+	argCnt := 4
+	valueStrings := make([]string, len(responses))
+	valueArgs := make([]interface{}, argCnt*len(responses))
+
+	for idx, resp := range responses {
+		valueStrings[idx] = "(?, ?, ?, ?)"
+
+		pointIdx := argCnt * idx
+
+		valueArgs[pointIdx] = resp.DotId
+		valueArgs[pointIdx+1] = resp.SpecId
+		valueArgs[pointIdx+2] = resp.Response
+		valueArgs[pointIdx+3] = resp.FinishedAt
```
Isn't there a way to pass a slice of structs to insert? Like with named arguments?

EasterTheBunny replied:
I found one with `sqlx` package. That's cleaner.

jmank88 commented on core/bridges/bridge_type.go:
```diff
@@ -199,3 +199,10 @@ func (t *BridgeName) Scan(value interface{}) error {
 }
+
+type BridgeResponse struct {
+	DotId      string
+	SpecId     int32
```
```suggestion
	DotID      string
	SpecID     int32
```

jmank88 commented on core/bridges/bridge_cache.go:
```diff
@@ -0,0 +1,290 @@
+	return nil
+}
+
+func (c *BridgeCache) ExternalInitiators(ctx context.Context, offset int, limit int) ([]ExternalInitiator, int, error) {
+	return c.orm.ExternalInitiators(ctx, offset, limit)
+}
+
+func (c *BridgeCache) CreateExternalInitiator(ctx context.Context, externalInitiator *ExternalInitiator) error {
+	return c.orm.CreateExternalInitiator(ctx, externalInitiator)
+}
+
+func (c *BridgeCache) DeleteExternalInitiator(ctx context.Context, name string) error {
+	return c.orm.DeleteExternalInitiator(ctx, name)
+}
+
+func (c *BridgeCache) FindExternalInitiator(ctx context.Context, eia *auth.Token) (*ExternalInitiator, error) {
+	return c.orm.FindExternalInitiator(ctx, eia)
+}
+
+func (c *BridgeCache) FindExternalInitiatorByName(ctx context.Context, iname string) (exi ExternalInitiator, err error) {
+	return c.orm.FindExternalInitiatorByName(ctx, iname)
+}
+
+func (c *BridgeCache) GetCachedResponseWithFinished(ctx context.Context, dotId string, specId int32, maxElapsed time.Duration) ([]byte, time.Time, error) {
+	return c.orm.GetCachedResponseWithFinished(ctx, dotId, specId, maxElapsed)
+}
+
+func (c *BridgeCache) BulkUpsertBridgeResponse(ctx context.Context, responses []BridgeResponse) error {
+	return c.orm.BulkUpsertBridgeResponse(ctx, responses)
+}
```
Is it possible to embed `ORM` directly so that these are automatically available, and we just override a subset?

EasterTheBunny committed:
addressed comments

jmank88 commented on core/bridges/bridge_cache.go:
```diff
@@ -0,0 +1,261 @@
+var _ ORM = (*BridgeCache)(nil)
+var _ services.Service = (*BridgeCache)(nil)
+
+func NewBridgeCache(base ORM, lggr logger.Logger, upsertInterval time.Duration) *BridgeCache {
```
Could shorten to just Cache since we are in package bridges already:
```suggestion
func NewCache(base ORM, lggr logger.Logger, upsertInterval time.Duration) *Cache {
```

EasterTheBunny committed:
rename to remove 'Bridge'

EasterTheBunny committed:
health report using CopyHealth

EasterTheBunny committed:
remove 'Bridge

EasterTheBunny committed:
remove bridge from test

EasterTheBunny committed:
add BridgeCache to health tests

EasterTheBunny committed:
fix data race

EasterTheBunny committed:
results of make generate

jmank88 commented on core/bridges/cache.go:
```diff
@@ -0,0 +1,260 @@
+		return nil
+	})
+}
+
+func (c *Cache) Ready() error {
+	return c.StateMachine.Ready()
+}
```
This should already happen automatically since it is embedded
```suggestion
```

jmank88 commented on core/bridges/cache.go:
```diff
@@ -0,0 +1,260 @@
+}
+
+func (c *Cache) Name() string {
+	return CacheServiceName
```
This method is meant to use the fully qualified name from the logger:
```suggestion
	return c.lggr.Name()
```

jmank88 commented on core/bridges/cache.go:
```diff
@@ -0,0 +1,260 @@
+}
+
+func (c *Cache) HealthReport() map[string]error {
+	return map[string]error{CacheServiceName: c.Healthy()}
```
We need the fully qualified name here too:
```suggestion
	return map[string]error{c.Name(): c.Healthy()}
```

jmank88 commented on core/bridges/cache.go:
```diff
@@ -0,0 +1,260 @@
+
+func (c *Cache) Close() error {
+	return c.StopOnce(CacheServiceName, func() error {
+		c.stop <- struct{}{}
```
nit, since there is only one thing listening now, but closing will have the same effect and prevent future errors if more goroutines are added:
```suggestion
		close(c.stop)
```

jmank88 commented on core/bridges/cache.go:
```diff
@@ -0,0 +1,260 @@
+
+func (c *Cache) doBulkUpsert() {
+	c.mu.RLock()
+	values := make([]BridgeResponse, 0, len(c.bridgeLastValueCache))
+
+	for _, value := range c.bridgeLastValueCache {
+		values = append(values, value)
+	}
+
```
```suggestion
	values := slices.Clone(c.bridgeLastValueCache)
```

EasterTheBunny replied:
This is a slice from a map, not a new slice from an existing slice. Is there a shortcut method for this?

jmank88 replied:
Ah, yeah there is: https://pkg.go.dev/golang.org/x/exp/maps#Values

EasterTheBunny committed:
addressed comments

EasterTheBunny committed:
simplify getting values from map

EasterTheBunny committed:
make linter happy

EasterTheBunny committed:
add logger to tests

EasterTheBunny committed:
fix test

samsondav commented on core/bridges/cache.go:
```diff
@@ -0,0 +1,253 @@
+	}
+
+	c.mu.Lock()
+	c.bridgeLastValueCache[responseKey(dotId, specId)] = BridgeResponse{
+		DotID:      dotId,
+		SpecID:     specId,
+		Value:      response,
+		FinishedAt: finishedAt,
+	}
```
Could this race and update a newer response with an older one?

samsondav commented on core/bridges/cache.go:
```diff
@@ -0,0 +1,253 @@
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	c.bridgeLastValueCache[responseKey(dotId, specId)] = BridgeResponse{
+		DotID:      dotId,
+		SpecID:     specId,
+		Value:      response,
+		FinishedAt: time.Now(),
+	}
```
Same question here about the race, also perhaps this function can be re-used above?

EasterTheBunny replied:
I moved the read and write lines to their own functions to consolidate the data access and locks. I also applied a time check on the upsert. This should handle the possibility of a data race.

samsondav commented on core/services/pipeline/runner.go:
```diff
@@ -124,21 +135,30 @@ func NewRunner(orm ORM, btORM bridges.ORM, cfg Config, bridgeCfg BridgeConfig, l
-func (r *runner) Start(context.Context) error {
+func (r *runner) Start(ctx context.Context) error {
 	return r.StartOnce("PipelineRunner", func() error {
+		// starting cache can happen here for ORM
```
What does this comment mean?

EasterTheBunny replied:
I clarified the comment and moved it to where it makes more sense.

samsondav commented on core/services/pipeline/runner.go:
```diff
@@ -124,21 +135,30 @@ func NewRunner(orm ORM, btORM bridges.ORM, cfg Config, bridgeCfg BridgeConfig, l
 			go r.runReaperLoop()
 		}
+
+		cache, isCache := r.btORM.(*bridges.Cache)
```
When would it not be a cache?

Could also check if it implements `Start` to make it more generally applicable

jmank88 commented on core/services/pipeline/runner.go:
```diff
@@ -108,10 +109,20 @@ var (
 	r := &runner{
 		orm:                    orm,
-		btORM:                  btORM,
+		btORM:                  bridges.NewCache(btORM, lggr, bridges.DefaultUpsertInterval),
```
Since the runner "owns" this cache, we should construct the runner's `Logger` before this, so that we can pass it in here. That keeps the service hierarchy organized instead of flat. i.e. `PipelineRunner.BridgeCache`

EasterTheBunny committed:
address comments

EasterTheBunny committed:
initialize logger before runner

EasterTheBunny committed:
cleanup tests

EasterTheBunny committed:
whitespace fix

samsondav reviewed (changes requested):
I think this can leak memory. Does it clean up expired values?

jmank88 commented on core/bridges/cache.go:
```diff
@@ -0,0 +1,269 @@
+
+const (
+	CacheServiceName         = "BridgeCache"
+	DefaultUpsertInterval    = 500 * time.Millisecond
+	DefaultBulkInsertTimeout = 250 * time.Millisecond
```
How were these values chosen?

EasterTheBunny replied:
Pin the tail on the donkey. Lol. Do you have some thoughts on what they should be? I was actually basing the timeout on being more than a standard query, though I don't know what qualifies as a long running query within this system. The upsert interval would need to be longer than the timeout and be regular enough to ensure we have a sensible backup, but not run so often than we overload the database again.

jmank88 replied:
Are we going to wish these were configurable later on? Maybe we should expose one of them, and then derive the other from that.

EasterTheBunny commented:
> I think this can leak memory. Does it clean up expired values?

It doesn't cleanup values and we made the decision not to do so in an external conversation due to the complexity of hooking into job removal.

EasterTheBunny committed:
increase interval and default to SQL timeout
